### PR TITLE
PUBDEV-6377 - Filling of sparse DMatrix data

### PR DESCRIPTION
**Pull request by Pscheidl**
[JIRA - PUBDEV-6377](https://0xdata.atlassian.net/browse/PUBDEV-6377)

Please also visit [Benchmarks](http://mr-0xc1:8080/view/H2O-3/job/h2o-3-benchmark-pipeline/job/pavel_pubdev-6377/) to see the performance impact of these changes.

### CODE REVIEW

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java [hunk 1]
```diff
@@ -522,76 +521,137 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
...
                     } else {
-                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, vecs[j].at8(i));
+                        colIndex[rowPointer][currentCol++] = di.getCategoricalId(j, vecs[j].at8(i));
                     }
                     nonZeroCount++;
                 }
 
                 for (int j = 0; j < di._nums; ++j) {
+                    if(currentCol == SPARSE_MATRIX_DIM){
+                        currentCol = 0;
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: I do believe a single check at each round will in the end (after all optimizations kick in) be a little bit faster than calculating the indices each time.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java [hunk 2]
```diff
@@ -522,76 +521,137 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
...
+                        rowPointer++;
+                    }
+                    data[rowPointer][currentCol] = 1; //one-hot encoding
+                    
+                    if(!foundNonZero){
+                        foundNonZero = true;
+                        for (int k = lastNonZeroRow; k < actualRows; k++) {
+                        rowHeaders[rowHeaderRowPointer][rowHeaderColPointer++] = nonZeroCount;
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: This doesn't seem correct to me. IMHO every time you increment a pointer you need to make sure there is room in the array.

Is there a mechanism that prevents you from running out of space in `rowHeaders[rowHeaderRowPointer]`

Comment (Pscheidl) [reply h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: There is a prevention mechanism. It is actually so simple it is easy to overlook. 

This particular "row header" (as we call it) has one purpose: to mark the first non-zero element each line begins with. As it required for this to happen exactly once per each row (if a row is empty, last non-zero is re-used, there is this special handling a little bit further in the code), the increment only happens once. Therefore, there is no need to check it more than once per each line.

This implies the actual check for `rowHeaderColPointer` overflow is NOT done in the ` for (int j = 0; j < di._cats; ++j) ` loop, but rather directly in the embedding loop `for(long i = f.anyVec().espc()[chunk]; i < f.anyVec().espc()[chunk+1]; i++)`. It is done right after weights are checked not to be null and the line verified for not to be skipped.

This piece of code does it:
```java
                if(rowHeaderColPointer == SPARSE_MATRIX_DIM){
                // enlarge final data arrays by 2x if needed
                    rowHeaderColPointer = 0;
                    rowHeaderRowPointer++;
                }
                boolean foundNonZero = false;
```

As seen in the example, there is also `foundNonZero` variable defined and explicitely set to false. This variable makes it possible to find out if there was a non-zero element found already on this row (as the outer loop is naturally traversing the matrix row-wise).

And as you can see a little bit further on line 550:

```java
                    if(!foundNonZero){
                        foundNonZero = true;
....
```

The non-zero element is only written into row headers if there was none discovered on this line before.

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostUtilsTest.java [hunk 3]
```diff
@@ -56,5 +174,338 @@ public void parseFeatureScores() throws IOException, ParseException {
...
+
+  private static CsrLayout createLayout(long size, int maxArrLen) {
+    CsrLayout l = new CsrLayout();
+    l._numRegRows = (int) (size / maxArrLen);
+    l._regRowLen = maxArrLen;
+    l._lastRowLen = (int) (size - ((long) l._numRegRows * l._regRowLen)); // allow empty last row (easier and it shouldn't matter)
+    return l;
+  }
+
+  private static class CsrLayout {
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostUtilsTest.java]: Borrowed from @michalkurka - Thank you !

**Comment by wendycwong** [comment]
Pavel:

It looks like the major change you made was the way you increment the row indices and this resolves the index out of bound problem?  Amazing.


Wendy

**Comment by Pscheidl** [comment]
@wendycwong Yes, unfortunately the CSR algorithm was implemented badly - there was no translation between 1D space and 2D internal representation.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java (hunk 2 above)
Comment (michalkurka) [reply h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: Thanks!

File: h2o-core/src/test/java/water/TestUtil.java [hunk 4]
```diff
@@ -1027,5 +1027,33 @@ protected static RowData toRowData(Frame fr, String[] columns, long row) {
...
     return result;
   }
+
+  /**
+   * @param f1 Frame to be comapared, not null
+   * @param f2 Frame to be compared, not null
+   * @return True if matrices are precisely the same - number of columns, rows & values at each cell.
+   * @throws IllegalStateException If any inequalities are found
+   */
+  public static boolean compareFrames(final Frame f1, final Frame f2) throws IllegalStateException {
```
Comment (michalkurka) [inline h2o-core/src/test/java/water/TestUtil.java]: Nice!

**Pull request merged**